### PR TITLE
Login redirect

### PR DESCRIPTION
**Pull request by arthurmcgregor**
- loginRedirect() will send user to the login page if they are not logged in and trying to view a page where they should be logged in (currently everything except /login / and /register)
- Will pass a query string next which will enable redirection to the correct page after successful login

😃

### CODE REVIEW

**Comment by arthurmcgregor** [comment]
Hmmmm so its failing the travis-ci check because the next variable is not defined. To pass it along from the route I took the same approach to what you have done with the api variable. However, I defined next in login.pug directly (whereas api is defined in base.pug)

So I'm not sure why next is failing the check when, in the same file a few lines earlier, api doesn't cause it to fail....

**Comment by CameronRP** [comment]
The api variable passed because of this line added to the top of user.js `/* global api, Promise */`
You should be able to add the line `/* global next, Promise */` to fix that issue.

**Comment by arthurmcgregor** [comment]
That sorted it - thanks :-)

File: public/javascripts/includes/user.js [hunk 1]
```diff
@@ -160,3 +164,20 @@ user.getHeaders = function() {
...
+  if (!user.isLoggedIn()) {
+    let next = window.location.pathname.slice(1);
+    window.location.href = '/login' + '?next=' + next;
+  }
+}
+
+// Do not run the loginRedirect on the public facing web pages
+if (window.location.pathname.slice(0,6) !== '/login' && window.location.pathname !== '/register' && window.location.pathname !== '/' ) {
+  loginRedirect();
+}
```
Comment (mjs) [inline public/javascripts/includes/user.js]: A few things:
- why use `slice()` just for `/login`?
- does the next URL need to be URI encoded?
- why do you remove the initial `/`?

Also, something like this is perhaps a little clearer (untested!):

```js

// These paths don't require that the user is logged in.
const noAuthPaths = Object.freeze([
  '/login', 
  '/register', 
  '/',
])

function loginRedirect(pathname) {
  if (noAuthPaths.includes(pathname) {
    return;
  }
  if (!user.isLoggedIn()) {
    window.location.href = '/login' + '?next=' + encodeURIComponent(pathname);
  }
}

loginRedirect(window.location.pathname);
```

**Review by mjs — COMMENTED** [review]
This is awesome. Thank you!

Just a few suggestions and comments.

**Comment by arthurmcgregor** [comment]
Thanks for the feedback @mjs. Have implemented your suggestions. The slicing was a hangover from before I realised there was an inbuilt process for handling query strings and wasn't needed anymore. URI encoding probably a good idea.

**Review by mjs — APPROVED** [review]
Awesome. I'm looking forward to having this change deployed. I was bitten by trying to use the UI without being logged in on Monday.

**Pull request merged**